x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/EnrichIT.java

```diff
@@ -0,0 +1,276 @@
+            while (rows.hasNext()) {
+                Iterator<Object> row = rows.next();
+                Object v = row.next();
+                Object k = row.next();
+                actualValues.put((String) k, (Double) v);
+            }
+            return actualValues;
+        };
+        try (var resp = run("from listens* | ENRICH songs | STATS avg(duration) by artist")) {
+            Map<String, Double> stats = extractStats.apply(resp);
+            assertThat(stats.keySet(), containsInAnyOrder("Eagles", "Linkin Park"));
+            assertThat(stats.get("Eagles"), closeTo(1.08333, 0.1));
+            assertThat(stats.get("Linkin Park"), closeTo(1.875, 0.1));
+        }
+    }
+
+    public void testListeningRatio() {
+        Function<EsqlQueryResponse, Map<String, Double>> extractStats = resp -> {
+            List<ColumnInfo> columns = resp.columns();
+            assertThat(columns, hasSize(2));
+            assertThat(columns.get(0).name(), equalTo("ratio"));
+            assertThat(columns.get(0).type(), equalTo("double"));
+            assertThat(columns.get(1).name(), equalTo("artist"));
+            assertThat(columns.get(1).type(), equalTo("keyword"));
+            Iterator<Iterator<Object>> rows = resp.values();
+            Map<String, Double> actualValues = new HashMap<>();
+            while (rows.hasNext()) {
+                Iterator<Object> row = rows.next();
+                Object v = row.next();
+                Object k = row.next();
+                actualValues.put((String) k, (Double) v);
+            }
+            return actualValues;
+        };
+
+        var statsCommand = " | ENRICH songs | STATS d = sum(duration), l = sum(length) by artist | EVAL ratio=d /l | KEEP ratio, artist";
+        try (var resp = run("from listens* " + statsCommand)) {
+            Map<String, Double> stats = extractStats.apply(resp);
+            assertThat(stats.keySet(), containsInAnyOrder("Eagles", "Linkin Park"));
+            assertThat(stats.get("Eagles"), closeTo(0.1521, 0.05));
+            assertThat(stats.get("Linkin Park"), closeTo(0.585, 0.05));
+        }
+    }
+
+    public void testFilterAfterEnrich() {
+        try (var resp = run("from listens* | ENRICH songs | WHERE length < 3.2 | limit 10 | KEEP artist,title")) {
+            Iterator<Object> row = resp.values().next();
+            assertThat(row.next(), equalTo("Linkin Park"));
+            assertThat(row.next(), equalTo("Numb"));
+        }
+    }
+
+    public void testTopN() {
+        try (var resp = run("from listens* | sort timestamp DESC | limit 1 | ENRICH songs | KEEP timestamp, artist")) {
+            Iterator<Object> row = resp.values().next();
+            assertThat(row.next(), equalTo(7L));
+            assertThat(row.next(), equalTo("Linkin Park"));
+        }
+        try (var resp = run("from listens* | ENRICH songs | sort timestamp DESC | limit 1 | KEEP timestamp, artist")) {
```

Comment:
nit: all the examples use only `ENRICH songs`; maybe it would be useful to also test `ENRICH songs ON ... WITH ...` (may require giving the enrich policy a second column to fully make sense).

Reason: while that's a bit on the paranoid side, there may be subtle differences, esp. the planner and optimizer might behave ever so slightly differently; we already had a bug because of that.

Reply:
ok, I pushed https://github.com/elastic/elasticsearch/pull/102099/commits/a5c06136164a5be9d4053f18c86149b72c4b8853.